lib/flight_scheduler/application.rb

```diff
@@ -85,9 +85,16 @@ def init_periodic_processor
           timer = Concurrent::TimerTask.new(**opts) do
             Async.logger.debug("Running periodic processor")
             job_registry.remove_old_jobs
+            job_registry.save
+            job_registry.jobs_in_state(Job::TERMINAL_STATES).each do |job|
+              Async.logger.info("Removing allocation for job in terminal state: id=#{job.display_id} state=#{job.state}")
+              FlightScheduler::Deallocation::Job.new(job).call
```

Comment:
Why is this step required? I assume it is about getting the job out of the `TERMINAL_STATES` and fully removed from the scheduler?

Does the daemon know it needs to respond `NODE_DEALLOCATED` to all of these messages, even if it has sent it previously?

Reply:
It a daemon is being restarted when the initial deallocation is being ran it won't have previously sent a `NODE_DEALLOCATED` message.  If that happens, we can end up with a job being stuck in a terminal state with nodes still allocated to it.

That can hold up other jobs from running through the scheduler.  Currently, the only solution to this was for a user to issue (another) cancellation request.

If the daemon has already received the deallocation request and responded with `NODE_DEALLOCATED` nothing breaks, the daemon will simply ignore the request.

Reply:
From reviewing the daemon, it looks like it responds to all `JOB_DEALLOCATED` requests with `NODE_DEALLOCATED` (if it can).

I  vaguely remember this now from when I implemented it. At the time I considered the double response on cancel a nuisance but otherwise harmless (as you said).

However as the controller now relies on the positive `NODE_DEALLOCATED` response, it is critical that the daemon continues to respond to `JOB_DEALLOCATED` requests it does not recognise.